File: programmers/python/level2/p42888.py
# 오픈 채팅방

'''
<방의 로그가 보여질 때 중간에 이름을 바꾸면 바꾼이름으로 로그가 보이게>
[parameter] 기록 리스트
[return] 최종 보여질 리스트

[풀이방법]
기록에 uid(바뀌지 않는 값), nickname(변경 가능), behavior(행동)
nicknames = {uid : nickname} 형식으로 저장해놓고
닉네임이 변할 수 있는 행동에서 닉네임을 업데이트 해줌
마지막 결과 리스트를 출력할 때 nicknames[uid]로 변환해서 로그 반환
'''

def solution(record):
    answer = []
    logs = []
    
    nicknames = {}
    
    for r in record:
        split_r = r.split(' ')
        
        behavior = split_r[0]
        uid = split_r[1]
        
        if behavior != 'Leave':
            nicknames[uid] = split_r[2]
        
        if behavior != 'Change':
            logs.append(behavior_to_kor(uid, behavior))
    
    for l in logs:
        answer.append(nicknames[l[0]]+l[1])
        
    return answer

def behavior_to_kor(uid, behavior):
    if behavior == 'Enter':
        return  [uid, '님이 들어왔습니다.']
    elif behavior == 'Leave':
        return  [uid, '님이 나갔습니다.']

print(solution())
        
    

File: book/java/src/y25/level1/P42888.java
package y25.level1;

import java.util.*;

public class P42888 {
    /*
    ["Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"]
    ["Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다."]
     */
    public String[] solution(String[] record) {
        Map<String, String> idMap = new HashMap<>();
        List<String[]> results = new ArrayList<>();

        for (String rc : record) {
            String[] split = rc.split(" ");
            if (split[0].equals("Leave")) {
                results.add(new String[]{split[1], "님이 나갔습니다."});
            } else {
                idMap.put(split[1], split[2]);
                if (split[0].equals("Enter")) {
                    results.add(new String[]{split[1], "님이 들어왔습니다."});
                }
            }
        }

        return results.stream()
                .map(result -> idMap.get(result[0]) + result[1])
                .toArray(String[]::new);
    }
}
